### DISCUSSION TITLE
fortls using too much RAM ?

### DISCUSSION THREAD

**Discussion by paulatz** [fortran-lang/fortls#435] [General]
Hello,
I use fortls through vsc for a large project that includes about 3200 fortran files, composed of several git repositories. I also have a separate directory which has compiled "historical" versions of the same code with about 45'000 fortran files in total, for quick reference.
I find that fortls is using too much memory, which can also spike sometime when I start a compilation, causing my system to hang.

I.e. here is a sample output for fortls running in backgroud while a single vsc windows is open but idle:
```
MiB Mem :  31333.5 total,   5396.8 free,  17154.9 used,   9402.2 buff/cache     
MiB Swap:      0.0 total,      0.0 free,      0.0 used.  14178.6 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                                              
  54962 paulatto  20   0 5827096   5.4g  12280 S   0.0  17.5   1:15.13 fortls                                                                                                                                                               
  54963 paulatto  20   0 5827096   5.4g  12288 S   0.0  17.5   1:14.52 fortls
```
I.e. there are two fortls processes, each is using 17.5% of my 32GB of RAM.

I have explicitly excluded the "historical" directory, but it does not seem to be the source of the problem. Is this memory usage normal? Or anybody has some hint on how to reduce it.
regards

**Comment by gnikit** [comment]
Without seeing the actual codebase, I think it will be hard to gauge if this is expected behaviour or not. ~5Gb of RAM is not much if you have a few tens of millions lines of code for 10M lines it averages to 500 bytes per line. Can you provide additional information, like number of lines of your project or even the codebase, your fortls settings, and anything else that could be relevant?

FYI fortls by default should be using 4 processes, unless you changed that, which could indicate that 2 files in your project are slow to parse.

**Comment by paulatz** [comment]
Sure, it is the Quantum ESPRESSO[1] git three, with my much smaller project nested within [2]. This totals about 1.36M lines (53MB of Fortran code). Which is still 1 order of magnitude smaller than your example. 
Regarding the threads, I understand that fortls uses 4 threads during scanning, which is normal and fine:
```
paulatto  295911  6.3  0.6 1214166820 198176 ?   Ssl  11:18   0:01  \_ /usr/share/code/code
paulatto  295914  0.0  0.1 34054176 48640 ?      S    11:18   0:00  |   \_ /usr/share/code/code --type=zygote --no-zygote-sandbox
paulatto  295953  3.9  0.5 34543280 167420 ?     Sl   11:18   0:00  |   |   \_ /usr/share/code/code --type=gpu-process --crashpad-handler-pid=295936 -
paulatto  295915  0.0  0.1 34054168 48128 ?      S    11:18   0:00  |   \_ /usr/share/code/code --type=zygote
paulatto  295917  0.0  0.0 34054196 12348 ?      S    11:18   0:00  |   |   \_ /usr/share/code/code --type=zygote
paulatto  296000 31.3  1.0 1216214156 349540 ?   Sl   11:18   0:07  |   |       \_ /usr/share/code/code --type=renderer --crashpad-handler-pid=295936 
paulatto  295957  0.4  0.2 33850416 70020 ?      Sl   11:18   0:00  |   \_ /usr/share/code/code --type=utility --utility-sub-type=network.mojom.Networ
paulatto  296573 10.8  0.4 1213945200 158928 ?   Sl   11:18   0:02  |   \_ /usr/share/code/code --type=utility --utility-sub-type=node.mojom.NodeServi
paulatto  296575 11.7  0.3 1213962152 120624 ?   Sl   11:18   0:02  |   \_ /usr/share/code/code --type=utility --utility-sub-type=node.mojom.NodeServi
paulatto  296824 25.9  0.8 1213947288 282528 ?   Sl   11:18   0:06  |   \_ /usr/share/code/code --type=utility --utility-sub-type=node.mojom.NodeServi
paulatto  297439 30.5  0.7 258012 225532 ?       Sl   11:18   0:07  |       \_ /home/paulatto/.vscode/extensions/ms-vscode.cpptools-1.22.11-linux-x64/
paulatto  297467 35.5  2.4 995180 791064 ?       Sl   11:18   0:08  |       \_ /usr/bin/python3 /home/paulatto/.local/bin/fortls --enable_code_actions
paulatto  297486 85.4  0.1  47684 34732 ?        R    11:18   0:18  |       |   \_ /usr/bin/python3 /home/paulatto/.local/bin/fortls --enable_code_act
paulatto  297487 85.6  0.1  48548 36092 ?        R    11:18   0:19  |       |   \_ /usr/bin/python3 /home/paulatto/.local/bin/fortls --enable_code_act
paulatto  297488 85.7  0.1  51524 39500 ?        R    11:18   0:19  |       |   \_ /usr/bin/python3 /home/paulatto/.local/bin/fortls --enable_code_act
paulatto  297489 86.0  0.1  48784 36284 ?        R    11:18   0:19  |       |   \_ /usr/bin/python3 /home/paulatto/.local/bin/fortls --enable_code_act
paulatto  297838  3.6  0.3 1211968152 116640 ?   Sl   11:18   0:00  |       \_ /usr/share/code/code /home/paulatto/.vscode/extensions/gitlab.gitlab-wo
paulatto  295936  0.0  0.0 33576916 3712 ?       Sl   11:18   0:00  \_ /usr/share/code/chrome_crashpad_handler --monitor-self-annotation=ptype=crashpa
```
The problem was that vsc was spawning two separate fortls process, with 4 threads each, which is not normal.

Anyway, since updating vsc to version 1.95 the other day, the problem seems to have solved itself and fortls is only using about 1GB virtual memory. I'll be back if the problems come back

Thank you

[1] https://gitlab.com/QEF/q-e
[2] https://github.com/anharmonic/d3q
